File: app/src/main/java/org/mifos/selfserviceapp/ui/fragments/ClientChargeFragment.java
package org.mifos.selfserviceapp.ui.fragments;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import org.mifos.selfserviceapp.R;
import org.mifos.selfserviceapp.models.Charge;
import org.mifos.selfserviceapp.presenters.ClientChargePresenter;
import org.mifos.selfserviceapp.ui.activities.base.BaseActivity;
import org.mifos.selfserviceapp.ui.adapters.ClientChargeAdapter;
import org.mifos.selfserviceapp.ui.fragments.base.BaseFragment;
import org.mifos.selfserviceapp.ui.enums.ChargeType;
import org.mifos.selfserviceapp.ui.views.ClientChargeView;
import org.mifos.selfserviceapp.utils.Constants;
import org.mifos.selfserviceapp.utils.RecyclerItemClickListener;
import org.mifos.selfserviceapp.utils.Toaster;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * @author Vishwajeet
 * @since 17/8/16.
 */

public class ClientChargeFragment extends BaseFragment implements
        RecyclerItemClickListener.OnItemClickListener, ClientChargeView {

    @Inject
    ClientChargePresenter mClientChargePresenter;

    @Inject
    ClientChargeAdapter clientChargeAdapter;

    @BindView(R.id.rv_client_charge)
    RecyclerView rvClientCharge;

    @BindView(R.id.swipe_charge_container)
    SwipeRefreshLayout swipeChargeContainer;

    @BindView(R.id.ll_error)
    RelativeLayout rlErrorLayout;

    @BindView(R.id.iv_status)
    ImageView ivError;

    @BindView(R.id.tv_status)
    TextView tvError;

    private long id;
    private ChargeType chargeType;
    private View rootView;
    private LinearLayoutManager layoutManager;
    private List<Charge> clientChargeList = new ArrayList<>();

    public static ClientChargeFragment newInstance(long clientId, ChargeType chargeType) {
        ClientChargeFragment clientChargeFragment = new ClientChargeFragment();
        Bundle args = new Bundle();
        args.putLong(Constants.CLIENT_ID, clientId);
        args.putSerializable(Constants.CHARGE_TYPE, chargeType);
        clientChargeFragment.setArguments(args);
        return clientChargeFragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ((BaseActivity) getActivity()).getActivityComponent().inject(this);
        if (getArguments() != null) {
            setToolbarTitle(getString(R.string.charges));
            id = getArguments().getLong(Constants.CLIENT_ID);
            chargeType = (ChargeType) getArguments().getSerializable(Constants.CHARGE_TYPE);
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
            @Nullable Bundle savedInstanceState) {
        rootView = inflater.inflate(R.layout.fragment_client_charge, container, false);
        ButterKnife.bind(this, rootView);

        mClientChargePresenter.attachView(this);

        layoutManager = new LinearLayoutManager(getActivity());
        layoutManager.setOrientation(LinearLayoutManager.VERTICAL);

        rvClientCharge.setLayoutManager(layoutManager);
        rvClientCharge.addOnItemTouchListener(new RecyclerItemClickListener(getActivity(), this));

        swipeChargeContainer.setColorSchemeResources(R.color.blue_light, R.color.green_light, R
                .color.orange_light, R.color.red_light);
        swipeChargeContainer.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                loadCharges();
            }
        });
        loadCharges();
        return rootView;
    }

    private void loadCharges() {
        rlErrorLayout.setVisibility(View.GONE);
        swipeChargeContainer.setVisibility(View.VISIBLE);

        if (chargeType == ChargeType.CLIENT) {
            mClientChargePresenter.loadClientCharges(id);
        } else if (chargeType == ChargeType.SAVINGS) {
            mClientChargePresenter.loadSavingsAccountCharges(id);
        } else if (chargeType == ChargeType.LOAN) {
            mClientChargePresenter.loadLoanAccountCharges(id);
        }
    }

    @Override
    public void showErrorFetchingClientCharges(String message) {
        Toaster.show(rootView, message);
    }

    @Override
    public void showClientCharges(List<Charge> clientChargeList) {
        this.clientChargeList = clientChargeList;
        inflateClientChargeList();
        if (swipeChargeContainer.isRefreshing()) {
            swipeChargeContainer.setRefreshing(false);
        }
    }

    private void inflateClientChargeList() {
        if (clientChargeList.size() > 0) {
            clientChargeAdapter.setClientChargeList(clientChargeList);
            rvClientCharge.setAdapter(clientChargeAdapter);
        } else {
            rlErrorLayout.setVisibility(View.VISIBLE);
            swipeChargeContainer.setVisibility(View.GONE);
            tvError.setText(getString(R.string.error_no_charge));
        }
    }

    @Override
    public void showProgress() {
        swipeChargeContainer.setRefreshing(true);
    }

    @Override
    public void hideProgress() {
        swipeChargeContainer.setRefreshing(false);
    }

    @Override
    public void onItemClick(View childView, int position) {

    }

    @Override
    public void onItemLongPress(View childView, int position) {

    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mClientChargePresenter.detachView();
    }

}

File: app/src/main/java/org/mifos/mobile/ui/client_charge/ClientChargeFragment.kt
package org.mifos.mobile.ui.client_charge

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.widget.LinearLayoutManager
import com.github.therajanmaurya.sweeterror.SweetUIErrorHandler
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import org.mifos.mobile.R
import org.mifos.mobile.databinding.FragmentClientChargeBinding
import org.mifos.mobile.models.Charge
import org.mifos.mobile.ui.activities.base.BaseActivity
import org.mifos.mobile.ui.adapters.ClientChargeAdapter
import org.mifos.mobile.ui.enums.ChargeType
import org.mifos.mobile.ui.fragments.base.BaseFragment
import org.mifos.mobile.utils.Constants
import org.mifos.mobile.utils.Network
import org.mifos.mobile.utils.ParcelableAndSerializableUtils.getCheckedArrayListFromParcelable
import org.mifos.mobile.utils.ParcelableAndSerializableUtils.getCheckedSerializable
import org.mifos.mobile.utils.Toaster

/*

@AndroidEntryPoint
class ClientChargeFragment : BaseFragment() {

    private var _binding: FragmentClientChargeBinding? = null
    private val binding get() = _binding!!

    private val viewModel: ClientChargeViewModel by viewModels()

    private var clientChargeAdapter: ClientChargeAdapter? = null
    private var id: Long? = 0
    private var chargeType: ChargeType? = null
    private var layoutManager: LinearLayoutManager? = null
    private var clientChargeList: List<Charge?>? = ArrayList()
    private var sweetUIErrorHandler: SweetUIErrorHandler? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        (activity as? BaseActivity)?.showToolbar()
        if (arguments != null) {
            id = arguments?.getLong(Constants.CLIENT_ID)
            chargeType = arguments?.getCheckedSerializable(
                ChargeType::class.java,
                Constants.CHARGE_TYPE
            ) as ChargeType
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        _binding = FragmentClientChargeBinding.inflate(inflater, container, false)
        clientChargeAdapter = ClientChargeAdapter(::onItemClick)
        setToolbarTitle(getString(R.string.charges))
        sweetUIErrorHandler = SweetUIErrorHandler(activity, binding.root)
        layoutManager = LinearLayoutManager(activity)
        layoutManager?.orientation = LinearLayoutManager.VERTICAL
        binding.rvClientCharge.layoutManager = layoutManager
        binding.swipeChargeContainer.setColorSchemeResources(
            R.color.blue_light,
            R.color.green_light,
            R.color.orange_light,
            R.color.red_light,
        )
        binding.swipeChargeContainer.setOnRefreshListener { loadCharges() }
        if (savedInstanceState == null) {
            loadCharges()
        }
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        viewLifecycleOwner.lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.clientChargeUiState.collect {
                    when (it) {
                        is ClientChargeUiState.Loading -> showProgress()
                        is ClientChargeUiState.Error -> {
                            hideProgress()
                           // showErrorFetchingClientCharges(getString(it.message))
                        }

                        is ClientChargeUiState.Success -> {
                            hideProgress()
                            showClientCharges(it.charges)
                        }
                    }
                }
            }
        }

        binding.layoutError.btnTryAgain.setOnClickListener {
            retryClicked()
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelableArrayList(
            Constants.CHARGES,
            clientChargeList?.let {
                ArrayList(
                    it,
                )
            },
        )
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        if (savedInstanceState != null) {
            val charges: List<Charge?> = savedInstanceState.getCheckedArrayListFromParcelable(
                    Charge::class.java,
                    Constants.CHARGES
                ) ?: listOf()
            showClientCharges(charges)
        }
    }

    /**
     * Fetches Charges for `id` according to `chargeType` provided.
     */
    private fun loadCharges() {
        if (binding.layoutError.root.visibility == View.VISIBLE) {
            sweetUIErrorHandler?.hideSweetErrorLayoutUI(
                binding.rvClientCharge,
                binding.layoutError.root
            )
        }
        when (chargeType) {
            ChargeType.CLIENT -> id?.let { viewModel.loadClientCharges(it) }
            ChargeType.SAVINGS -> id?.let { viewModel.loadSavingsAccountCharges(it) }
            ChargeType.LOAN -> id?.let { viewModel.loadLoanAccountCharges(it) }
            else -> {}
        }
    }

    /**
     * It is called whenever any error occurs while executing a request. If not connected to
     * internet then it shows display a message to user to connect to internet other it just
     * displays the `message` in a {@link Snackbar}
     *
     * @param message Error message that tells the user about the problem.
     */
    fun showErrorFetchingClientCharges(message: String?) {
        if (!Network.isConnected(activity)) {
            sweetUIErrorHandler?.showSweetNoInternetUI(
                binding.rvClientCharge,
                binding.layoutError.root
            )
        } else {
            sweetUIErrorHandler?.showSweetErrorUI(
                message,
                binding.rvClientCharge,
                binding.layoutError.root
            )
            Toaster.show(binding.root, message)
        }
    }

    /**
     * Tries to fetch charges again.
     */
    fun retryClicked() {
        if (Network.isConnected(context)) {
            sweetUIErrorHandler?.hideSweetErrorLayoutUI(
                binding.rvClientCharge,
                binding.layoutError.root
            )
            loadCharges()
        } else {
            Toast.makeText(
                context,
                getString(R.string.internet_not_connected),
                Toast.LENGTH_SHORT,
            ).show()
        }
    }

    fun showClientCharges(clientChargeList: List<Charge?>?) {
        this.clientChargeList = clientChargeList
        inflateClientChargeList()
        if (binding.swipeChargeContainer.isRefreshing) {
            binding.swipeChargeContainer.isRefreshing = false
        }
    }

    /**
     * Updates `clientChargeAdapter` with updated `clientChargeList` if
     * `clientChargeList` size if greater than 0 else shows the error layout
     */
    private fun inflateClientChargeList() {
        if (clientChargeList?.isNotEmpty() == true) {
            clientChargeAdapter?.setClientChargeList(clientChargeList)
            binding.rvClientCharge.adapter = clientChargeAdapter
        } else {
            sweetUIErrorHandler?.showSweetEmptyUI(
                getString(R.string.charges),
                R.drawable.ic_charges,
                binding.rvClientCharge,
                binding.layoutError.root
            )
        }
    }

    fun showProgress() {
        binding.swipeChargeContainer.isRefreshing = true
    }

    fun hideProgress() {
        binding.swipeChargeContainer.isRefreshing = false
    }

    fun onItemClick(position: Int) {}
    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    companion object {
        fun newInstance(clientId: Long?, chargeType: ChargeType?): ClientChargeFragment {
            val clientChargeFragment = ClientChargeFragment()
            val args = Bundle()
            if (clientId != null) {
                args.putLong(Constants.CLIENT_ID, clientId)
            }
            args.putSerializable(Constants.CHARGE_TYPE, chargeType)
            clientChargeFragment.arguments = args
            return clientChargeFragment
        }
    }
}

 */